File: cpp/organism/animal/Animal.cpp
//
// Created by Filip on 12/04/2025.
//

#include "Animal.h"
#include "../../world/World.h"
#include <random>
#include <utility>

Animal::Animal(Point point, std::string symbol, int strength, int initiative,
               World &world)
    : Organism(point, std::move(symbol), strength, initiative, world) {}

bool Animal::collision(Organism &other) {
  if (this->getSymbol() == other.getSymbol()) {
    // Rozmnażanie, jeśli organizmy są tego samego gatunku
    reproduce();
  } else {
    // Logika walki, jeśli organizmy są różnych gatunków
    return !fight(other);
  }
  return false;
}

bool Animal::fight(Organism &other) {
  // Pusta definicja metody
  if (this->getStrength() < other.getStrength()) {
    world.getLogger().logEvent(
        "Organizm " + this->getSymbol() + " został pokonany przez " +
        other.getSymbol());
    die();
    return true;
  }

  world.getLogger().logEvent(
      "Organizm " + this->getSymbol() + " pokonał " + other.getSymbol());
  other.die();
  return false;

}

File: java/src/organism/animal/Animal.java
package organism.animal;

import logger.EventLogger;
import organism.Organism;
import world.World;

import java.net.URL;

public abstract class Animal extends Organism {

    public Animal(int x, int y, String symbol, int strength, int initiative, World world) {
        super(x, y, symbol, strength, initiative, world);
    }

    @Override
    public boolean collision(Organism other) {
        if (other.getAge() <= 0) {
            return false; // No collision
        }
        if (this.getSymbol() == other.getSymbol()) {
            // Reproduction if same species
            reproduce();
            return false;
        } else {
            // Fight if different species
            return !fight(other);
        }
    }

    protected boolean fight(Organism other) {
        EventLogger.getInstance().log("Fight between " + this.getClass().getName() + " and " + other.getClass().getName());
        if (this.getStrength() < other.getStrength()) {
            // Log defeat
            EventLogger.getInstance().log("Organism " + this.getSymbol() + " was defeated by " + other.getSymbol());
            this.die();
            return true; // This organism died
        }
        EventLogger.getInstance().log("Organism " + other.getClass().getSimpleName() + " was defeated by " + other.getClass().getName() );
        other.die();
        return false; // This organism survived
    }

    protected abstract void reproduce();
}